## Instalação de Dependências

O código abaixo instala as bibliotecas [pandas](https://pandas.pydata.org/) e [gdown](https://pypi.org/project/gdown/) na sua máquina. O `gdown` é opcional e serve somente para baixar o dataset do google drive para seu diretório local. 

In [24]:
!pip install pandas gdown

Defaulting to user installation because normal site-packages is not writeable


## Importações

Nesse exemplo, usaremos apenas a biblioteca **pandas**. No código, é muito comum que precisemos digitar repetidamente a palavra `pandas` para acessar alguma função. Por exemplo:

```python
    pandas.read_csv('my_dataset.csv')
    pandas.pd.DataFrame(data)
    pd.to_csv('processed_data.csv')
```

Então, costuma-se abreviar `pandas` por `pd` no ato da importação

In [22]:
import pandas as pd

## Baixando Dataset

O dataset desse notebook está guardado num diretório do Google Drive. Para baixá-lo, execute a célula abaixo

In [27]:
import gdown
url = 'https://drive.google.com/uc?id=1ZfQJJTPNv2cY_cWKTnNbYW511l0oStJy'
save_path = 'SISU_USP_2022.csv'

# baixa o dataset em seu diretório local
try:
    print('Baixando dataset do Google Drive...')
    gdown.download(url, save_path, quiet=False)
    print('sucesso!')
except Exception as error:
    print(f'Erro ao baixar dataset\n{error}')

Baixando dataset do Google Drive...


Downloading...
From: https://drive.google.com/uc?id=1ZfQJJTPNv2cY_cWKTnNbYW511l0oStJy
To: /home/henrique/Projetos/Python/thm-ai-crash-course/content/pandas_basics/SISU_USP_2022.csv
100%|██████████| 10.1M/10.1M [00:02<00:00, 3.45MB/s]

sucesso!


## Carregamento do Dataset

Nesse exemplo, trabalhamos com o dataset do [Relatório SISU - Chamada Regular 1/2022](https://dadosabertos.mec.gov.br/sisu), especificamente para a Universidade de São Paulo (USP).

Inicialmente, precisamos carregar o dataset no notebook. Para isso, usamos a função `read_csv()` do pandas. Você deve passar como argumento o caminho do dataset. No nosso caso, ele está no mesmo diretório que o notebook. Então, basta apenas informar o nome do arquivo `'SISU_USP_2022.csv'`. Por exemplo:

```python
    dataframe = pd.read_csv('SISU_USP_2022.csv')
```

In [41]:
dataframe = pd.read_csv('SISU_USP_2022.csv')

O resultado dessa operação nada mais é que um dataframe do pandas, um objeto da biblioteca que representa um dataset. Tradicionalmente, abrevia-se dataframe por df. Então, a forma mais comum de se encontrar uma leitura de dataset é no formato:
```python
    df = pd.read_csv('dataframe.csv')
```

### Informações básicas do dataset

Inicialmente, podemos usar a função ```info()``` do pandas, para verificar um resumo informativo do dataset.

In [42]:
dataframe.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19716 entries, 0 to 19715
Data columns (total 50 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   ETAPA                  19716 non-null  int64  
 1   DS_ETAPA               19716 non-null  object 
 2   CODIGO_CAMPUS          19716 non-null  int64  
 3   NOME_CAMPUS            19716 non-null  object 
 4   UF_CAMPUS              19716 non-null  object 
 5   MUNICIPIO_CAMPUS       19716 non-null  object 
 6   CODIGO_CURSO           19716 non-null  int64  
 7   NOME_CURSO             19716 non-null  object 
 8   GRAU                   19716 non-null  object 
 9   TURNO                  19716 non-null  object 
 10  DS_PERIODICIDADE       19716 non-null  object 
 11  TP_COTA                0 non-null      float64
 12  TIPO_MOD_CONCORRENCIA  19716 non-null  object 
 13  MOD_CONCORRENCIA       19716 non-null  object 
 14  QT_VAGAS_CONCORRENCIA  19716 non-null  int64  
 15  PE

Como podemos observar, o dataset é gigantesco. Ele possui 50 colunas e 19716 linhas, sendo cada coluna representante de uma propriedade ou atributo do dataset, nesse caso ano da edição, nome da faculdade, nota do aluno etc., e cada linha um aluno. podemos verificar o formato do dataset, ou seja linhas x colunas, pela propriedade `shape`

In [5]:
dataframe.shape

(19716, 48)

### resumo estatístico

No dataset, existem algumas colunas numéricas, como as colunas de notas do aluno. Podemos estudar dados estatísticos básicos do **dataframe** por meio da função `describe()` do **pandas**. Ela retorna um dataframe com propriedades tais quais contagem, média, mínimo, máximo, desvio padrão, quartis etc.

In [30]:
dataframe.describe()

ETAPA  CODIGO_CAMPUS  CODIGO_CURSO  TP_COTA  QT_VAGAS_CONCORRENCIA  \
count  19716.0   1.971600e+04  1.971600e+04      0.0           19716.000000   
mean       7.0   8.065448e+05  3.744395e+05      NaN              10.836529   
std        0.0   4.664224e+05  1.084086e+06      NaN              12.415828   
min        7.0   1.298000e+03  2.859000e+03      NaN               1.000000   
25%        7.0   8.066000e+03  2.916000e+03      NaN               4.000000   
50%        7.0   1.075933e+06  4.660800e+04      NaN               7.000000   
75%        7.0   1.079570e+06  9.162500e+04      NaN              13.000000   
max        7.0   1.079833e+06  5.001352e+06      NaN              73.000000   

       PERCENTUAL_BONUS       PESO_CN  NOTA_MINIMA_L  NOTA_MINIMA_CH  \
count               0.0  19716.000000   19716.000000    19716.000000   
mean                NaN      1.702399     387.815987      384.455772   
std                 NaN      0.890849     157.588915      157.064283   
min                 NaN      1.000000       0.000000        0.000000   
25%                 NaN      1.000000     350.000000      350.000000   
50%                 NaN      1.000000     450.000000      400.000000   
75%                 NaN      2.000000     500.000000      500.000000   
max                 NaN      5.000000     600.000000      600.000000   

       NOTA_MINIMA_CN  NOTA_MINIMA_M  MEDIA_MINIMA         OPCAO  \
count     19716.00000   19716.000000   19716.00000  19716.000000   
mean        384.86407     394.561777     390.26618      1.267752   
std         157.31025     162.830080     158.59366      0.442799   
min           0.00000       0.000000       0.00000      1.000000   
25%         350.00000     350.000000     360.00000      1.000000   
50%         400.00000     450.000000     430.00000      1.000000   
75%         500.00000     500.000000     500.00000      2.000000   
max         600.00000     750.000000     630.00000      2.000000   

             NOTA_R  NOTA_R_COM_PESO  CLASSIFICACAO  
count  19716.000000     19716.000000   19716.000000  
mean     706.281193      1599.572428     181.421434  
std      138.500307       711.219999     232.419831  
min      260.000000       260.000000       1.000000  
25%      600.000000      1120.000000      34.000000  
50%      700.000000      1560.000000      92.000000  
75%      820.000000      1980.000000     236.000000  
max      980.000000      4900.000000    1286.000000

## Explorando o Dataset

### Explorando colunas

Podemos verificar quais as colunas do dataset por meio da propriedade `columns` do dataframe.

In [31]:
dataframe.columns

Index(['ETAPA', 'DS_ETAPA', 'CODIGO_CAMPUS', 'NOME_CAMPUS', 'UF_CAMPUS',
       'MUNICIPIO_CAMPUS', 'CODIGO_CURSO', 'NOME_CURSO', 'GRAU', 'TURNO',
       'DS_PERIODICIDADE', 'TP_COTA', 'TIPO_MOD_CONCORRENCIA',
       'MOD_CONCORRENCIA', 'QT_VAGAS_CONCORRENCIA', 'PERCENTUAL_BONUS',
       'PESO_L', 'PESO_CH', 'PESO_CN', 'PESO_M', 'PESO_R', 'NOTA_MINIMA_L',
       'NOTA_MINIMA_CH', 'NOTA_MINIMA_CN', 'NOTA_MINIMA_M', 'NOTA_MINIMA_R',
       'MEDIA_MINIMA', 'CPF', 'INSCRICAO_ENEM', 'INSCRITO', 'SEXO',
       'DATA_NASCIMENTO', 'UF_CANDIDATO', 'MUNICIPIO_CANDIDATO', 'OPCAO',
       'NOTA_L', 'NOTA_CH', 'NOTA_CN', 'NOTA_M', 'NOTA_R', 'NOTA_L_COM_PESO',
       'NOTA_CH_COM_PESO', 'NOTA_CN_COM_PESO', 'NOTA_M_COM_PESO',
       'NOTA_R_COM_PESO', 'NOTA_CANDIDATO', 'NOTA_CORTE', 'CLASSIFICACAO',
       'APROVADO', 'MATRICULA'],
      dtype='object')

Para acessar uma coluna em especial, você pode selecioná-la a partir de seu nome, pela sintaxe 
```python
    dataframe['column_name']
```
sendo dataframe o dataset do **pandas** e ```'column_name'``` o nome da coluna, conforme os exemplos abaixo:

Como primeiro exemplo, vejamos a coluna ```SEXO``` do dataframe

In [32]:
dataframe['SEXO']

0        M
1        F
2        F
3        M
4        M
        ..
19711    F
19712    F
19713    F
19714    F
19715    M
Name: SEXO, Length: 19716, dtype: object

Aparentemente, só existem dois valores possíveis nessa coluna:

- **M**: representando sexo masculino
- **F**: representando sexo feminino

Mas, como podemos ter certeza? Uma forma de descobrir é por meio da função ```unique()```

In [33]:
dataframe['SEXO'].unique()

array(['M', 'F'], dtype=object)

Conforme podemos observar, de fato, a coluna sexo possui somente **M** e **F** como valores.

Um outro exemplo mais interessante é verificar a coluna ```UF_CANDIDATO```, que contém a unidade federativa do aluno.

In [34]:
dataframe['UF_CANDIDATO'].unique()

array(['MG', 'SP', 'RO', 'MA', 'BA', 'AL', 'SC', 'MS', 'PE', 'PA', 'MT',
       'DF', 'CE', 'GO', 'TO', 'PB', 'RJ', 'PI', 'AM', 'RS', 'RN', 'PR',
       'ES', 'SE', 'AP', 'AC', 'RR'], dtype=object)

Conforme podemos observar, a coluna possui todas as siglas das unidades federativas do Brasil. Ou seja, o SISU recebeu inscrições de candidatos de todos os estados brasileiros.

Outra função interessante é que podemos selecionar mais de uma coluna por vez. Por exemplo, as colunas 'NOTA_L', 'NOTA_CH', 'NOTA_CN', 'NOTA_M' e 'NOTA_R' correspondem às notas dos alunos nos eixos de conhecimentos, isto é,  linguagens, ciências humanas, ciências da natureza, matemática e redação. Para selecionar essas colunas, podemos passar uma lista de nomes de colunas:

```python
    dataframe[['column_name_1', 'column_name_2', 'column_name_3', ... ,'column_name_n']]
```

Note que temos dois colchetes ao invés de um só. Isso ocorre porque estamos passando uma lista ao invés de um valor único. A sintaxe acima, por exemplo, é idêntica a:


```python
    column_names = ['column_name_1', 'column_name_2', 'column_name_3', ... ,'column_name_n']
    dataframe[column_names]
```

In [35]:
dataframe[['NOTA_L', 'NOTA_CH', 'NOTA_CN', 'NOTA_M', 'NOTA_R']]

NOTA_L NOTA_CH NOTA_CN NOTA_M  NOTA_R
0      565,9   663,2   618,5  702,8     780
1      459,2   503,5   459,7  533,8     460
2      502,9     529   519,9  617,4     600
3      527,2   562,3   481,4  461,9     560
4      582,7   629,5   522,3  667,2     860
...      ...     ...     ...    ...     ...
19711  636,4   654,3     621  729,5     800
19712  607,4   608,9   592,1  634,8     680
19713  597,3   628,5   564,4    594     640
19714  568,7   559,1   532,1  621,7     840
19715  568,6   601,6   612,5  628,7     680

[19716 rows x 5 columns]

### Dividindo o dataset

Por enquanto, temos um conjunto gigantesco de dados que está todo unido. No entanto, podemos separá-lo em diversos pedaços, segundo algum critério específico, para conduzir uma série de análises estatísticas e investigações. Por exemplo, que tal separar o dataset por unidade federativa?

Vamos começar separando um estado em específico. Por exemplo, vamos criar um dataset somente de alunos paulistas. Para isso, o pandas oferece uma sintaxe interessante de seleção:

```python
    dataframe[dataframe['column_name'] <condition> ]
```

Sendo `column_name` o nome da coluna e <condition> uma condição que ela deve satisfazer. Para separar São Paulo do resto do Brasil, poderíamos colocar a condição de que a propriedade `UF_IES` seja igual a `'SP'`.

In [36]:
sp_dataframe = dataframe[dataframe['UF_CANDIDATO'] == 'SP']

sp_dataframe

ETAPA         DS_ETAPA  CODIGO_CAMPUS  \
1          7  LISTA DE ESPERA           4255   
2          7  LISTA DE ESPERA           4255   
3          7  LISTA DE ESPERA           4255   
4          7  LISTA DE ESPERA           4255   
5          7  LISTA DE ESPERA           4255   
...      ...              ...            ...   
19711      7  LISTA DE ESPERA        1075932   
19712      7  LISTA DE ESPERA        1075932   
19713      7  LISTA DE ESPERA        1075932   
19714      7  LISTA DE ESPERA        1075932   
19715      7  LISTA DE ESPERA        1075932   

                                           NOME_CAMPUS UF_CAMPUS  \
1      Escola de Artes, Ciências e Humanidades - EACH        SP   
2      Escola de Artes, Ciências e Humanidades - EACH        SP   
3      Escola de Artes, Ciências e Humanidades - EACH        SP   
4      Escola de Artes, Ciências e Humanidades - EACH        SP   
5      Escola de Artes, Ciências e Humanidades - EACH        SP   
...                                                ...       ...   
19711                    Instituto Oceanográfico - IO        SP   
19712                    Instituto Oceanográfico - IO        SP   
19713                    Instituto Oceanográfico - IO        SP   
19714                    Instituto Oceanográfico - IO        SP   
19715                    Instituto Oceanográfico - IO        SP   

      MUNICIPIO_CAMPUS  CODIGO_CURSO     NOME_CURSO         GRAU     TURNO  \
1           São Paulo       1404700  BIOTECNOLOGIA  Bacharelado  Matutino   
2           São Paulo       1404700  BIOTECNOLOGIA  Bacharelado  Matutino   
3           São Paulo       1404700  BIOTECNOLOGIA  Bacharelado  Matutino   
4           São Paulo       1404700  BIOTECNOLOGIA  Bacharelado  Matutino   
5           São Paulo       1404700  BIOTECNOLOGIA  Bacharelado  Matutino   
...                ...           ...            ...          ...       ...   
19711       São Paulo         51975   OCEANOGRAFIA  Bacharelado  Integral   
19712       São Paulo         51975   OCEANOGRAFIA  Bacharelado  Integral   
19713       São Paulo         51975   OCEANOGRAFIA  Bacharelado  Integral   
19714       São Paulo         51975   OCEANOGRAFIA  Bacharelado  Integral   
19715       São Paulo         51975   OCEANOGRAFIA  Bacharelado  Integral   

       ... NOTA_L_COM_PESO  NOTA_CH_COM_PESO NOTA_CN_COM_PESO NOTA_M_COM_PESO  \
1      ...           459,2             503,5            919,4          1067,6   
2      ...           502,9               529           1039,8          1234,8   
3      ...           527,2             562,3            962,8           923,8   
4      ...           582,7             629,5           1044,6          1334,4   
5      ...           578,7             554,1            979,4          1040,4   
...    ...             ...               ...              ...             ...   
19711  ...          1272,8             654,3             1863          2188,5   
19712  ...          1214,8             608,9           1776,3          1904,4   
19713  ...          1194,6             628,5           1693,2            1782   
19714  ...          1137,4             559,1           1596,3          1865,1   
19715  ...          1137,2             601,6           1837,5          1886,1   

       NOTA_R_COM_PESO  NOTA_CANDIDATO NOTA_CORTE CLASSIFICACAO  APROVADO  \
1                  460           487,1     640,59           155         S   
2                  600          558,07     640,59            93         S   
3                  560          505,16     640,59           148         S   
4                  860          635,89     640,59            22         S   
5                  600          536,09     640,59           119         S   
...                ...             ...        ...           ...       ...   
19711             2400          698,22      657,6             4         S   
19712             2040           628,7      657,6            21         S   
19713     

Da mesma forma, podemos separar o Rio de Janeiro

In [37]:
rj_dataframe = dataframe[dataframe['UF_CANDIDATO'] == 'RJ']

rj_dataframe

ETAPA         DS_ETAPA  CODIGO_CAMPUS  \
459        7  LISTA DE ESPERA           4255   
495        7  LISTA DE ESPERA           4255   
973        7  LISTA DE ESPERA           4255   
1304       7  LISTA DE ESPERA           4255   
1595       7  LISTA DE ESPERA           4255   
...      ...              ...            ...   
19509      7  LISTA DE ESPERA        1075953   
19576      7  LISTA DE ESPERA        1075953   
19593      7  LISTA DE ESPERA        1075953   
19685      7  LISTA DE ESPERA        1075932   
19694      7  LISTA DE ESPERA        1075932   

                                           NOME_CAMPUS UF_CAMPUS  \
459    Escola de Artes, Ciências e Humanidades - EACH        SP   
495    Escola de Artes, Ciências e Humanidades - EACH        SP   
973    Escola de Artes, Ciências e Humanidades - EACH        SP   
1304   Escola de Artes, Ciências e Humanidades - EACH        SP   
1595   Escola de Artes, Ciências e Humanidades - EACH        SP   
...                                                ...       ...   
19509     Instituto de Relações Internacionais - IRI        SP   
19576     Instituto de Relações Internacionais - IRI        SP   
19593     Instituto de Relações Internacionais - IRI        SP   
19685                    Instituto Oceanográfico - IO        SP   
19694                    Instituto Oceanográfico - IO        SP   

      MUNICIPIO_CAMPUS  CODIGO_CURSO                       NOME_CURSO  \
459         São Paulo         91605      EDUCAÇÃO FÍSICA E SAÚDE   
495         São Paulo         91605      EDUCAÇÃO FÍSICA E SAÚDE   
973         São Paulo         91607                GESTÃO AMBIENTAL   
1304        São Paulo         91625  GESTÃO DE POLÍTICAS PÚBLICAS   
1595        São Paulo         91609                  LAZER E TURISMO   
...                ...           ...                              ...   
19509       São Paulo         60200        RELAÇÕES INTERNACIONAIS   
19576       São Paulo         60200        RELAÇÕES INTERNACIONAIS   
19593       São Paulo         60200        RELAÇÕES INTERNACIONAIS   
19685       São Paulo         51975                     OCEANOGRAFIA   
19694       São Paulo         51975                     OCEANOGRAFIA   

              GRAU     TURNO  ... NOTA_L_COM_PESO  NOTA_CH_COM_PESO  \
459    Bacharelado  Matutino  ...           592,4             712,9   
495    Bacharelado  Matutino  ...           520,8             697,2   
973    Bacharelado   Noturno  ...           536,8             642,2   
1304   Bacharelado   Noturno  ...          1230,6            1381,8   
1595   Bacharelado   Noturno  ...           971,8            1532,7   
...            ...       ...  ...             ...               ...   
19509  Bacharelado  Matutino  ...          1282,2              1403   
19576  Bacharelado  Matutino  ...          1123,8            1279,6   
19593  Bacharelado   Noturno  ...          1178,8            1299,8   
19685  Bacharelado  Integral  ...          1201,8             700,5   
19694  Bacharelado  Integral  ...          1098,2             583,5   

      NOTA_CN_COM_PESO NOTA_M_COM_PESO  NOTA_R_COM_PESO  NOTA_CANDIDATO  \
459               1243           682,9             1600          690,17   
495             1216,4           673,4             1800          701,11   
973             1296,4           704,6             1640          688,57   
1304             582,1          1291,4             2700          718,59   
1595             502,5            1157             1800          542,18   
...                ...             ...              ...             ...   
19509            567,3          1370,4             1560          686,99   
19576            570,5            1260             1680           657,1   
19593            615,9            1382             1760          692,94   
19685             1848          2257,5             2700          725,65   
19694             1767          1677,6             1920   

Agora, vamos por exemplo verificar as médias de notas de cada um desses dois estados e compará-las

In [47]:
sp_NOTA_R = sp_dataframe['NOTA_R'].dropna().mean()
rj_NOTA_R = rj_dataframe['NOTA_R'].dropna().mean()

print(f'Média da nota de Redação em São Paulo: {sp_NOTA_R}')
print(f'Média da nota de Redação em Rio de Janeiro: {rj_NOTA_R}')

Média da nota de Redação em São Paulo: 698.3117759532112
Média da nota de Redação em Rio de Janeiro: 783.9455782312925


### Criando novas colunas

No nosso dataset, existem duas colunas importantes: `'NOTA_CANDIDATO'` e `'NOTA_CORTE'`. Se um candidato tiver nota maior que a nota de corte, ele tem chances de ser aprovado. Então, que tal criar uma coluna nova no dataset que indique isso? Para criar uma nova coluna no dataset, utilize a seguinte sintaxe:

```python
    dataframe['new_column_name'] = <new_column>
```

Em que `'new_column_name'` é o nome da nova coluna e `new_column` é a nova coluna, uma `Series` do **pandas**

Vamos primeira criar uma nova coluna denominada `'SUP_CORTE'` (superou nota de corte) que é do tipo `bool` e é verdadeira se a nota do candidato é maior que a nota de corte.

In [17]:
sup_corte = dataframe['NOTA_CANDIDATO'] > dataframe['NOTA_CORTE']
sup_corte

0         True
1        False
2        False
3        False
4        False
         ...  
19711     True
19712    False
19713    False
19714    False
19715    False
Length: 19716, dtype: bool

Agora, vamos inserir essa nova coluna no dataset

In [19]:
dataframe['SUP_CORTE'] = sup_corte

Para verificar que a nova coluna está de fato no dataset, podemos tentar acessá-la

In [20]:
dataframe['SUP_CORTE']

0         True
1        False
2        False
3        False
4        False
         ...  
19711     True
19712    False
19713    False
19714    False
19715    False
Name: SUP_CORTE, Length: 19716, dtype: bool

Como podemos ver, a nova coluna está no dataset